presto-main/src/main/java/io/prestosql/cost/TopNStatsRule.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.cost;
+
+import io.prestosql.Session;
+import io.prestosql.matching.Pattern;
+import io.prestosql.spi.block.SortOrder;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.TypeProvider;
+import io.prestosql.sql.planner.iterative.Lookup;
+import io.prestosql.sql.planner.plan.TopNNode;
+
+import java.util.Optional;
+
+import static io.prestosql.sql.planner.plan.Patterns.topN;
+
+public class TopNStatsRule
+        extends SimpleStatsRule<TopNNode>
+{
+    private static final Pattern<TopNNode> PATTERN = topN();
+
+    public TopNStatsRule(StatsNormalizer normalizer)
+    {
+        super(normalizer);
+    }
+
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    protected Optional<PlanNodeStatsEstimate> doCalculate(TopNNode node, StatsProvider statsProvider, Lookup lookup, Session session, TypeProvider types)
+    {
+        PlanNodeStatsEstimate sourceStats = statsProvider.getStats(node.getSource());
+        double rowCount = sourceStats.getOutputRowCount();
+
+        if ((rowCount <= node.getCount()) || (node.getStep() == TopNNode.Step.PARTIAL)) {
```

Comment:
Not sure about `PARTIAL` handling. It does not limit output size to same extent as `FINAL`, yet it still can very substantially limit it vs size of the input to `queryParalellism*limit`. @findepi do you have an idea how to do better estimation here?

Reply:
There's no good answer for PARTIAL. The number of global rows coming out of that operation will depend on the number of tasks that get created at runtime, and that's not known in advance.

Reply:
See also https://github.com/prestosql/presto/blob/c74b775e103a7a239aafef1b7ca15fe6370495fe/presto-main/src/main/java/io/prestosql/cost/AggregationStatsRule.java#L56-L58

Reply:
After discussing with @findepi  i think we should go same approach here:
```java
if (node.getStep() != SINGLE) {
    return Optional.empty();
}
```

It should still help as join enumeration, which uses stats is happening before `SINGLE` `TopNNodes` are split into `PARTIAL` and `FINAL` (unless I am misreading the code). (*)


(*) It looks like the only way `TopNNode` can be converted into `PARTIAL`, before join enumeration, is due to execution of `PushLimitThroughOuterJoin` or `PushLimitThroughUnion`, which enforce `partial` flag on `LimitNode` they produce. Not sure if that logic is needed - I will look more into that.

Reply:
just wondering why `Optional.empty()` and not `Optional.of(sourceStats);`?

Reply:
`Optional.empty()` means we cannot compute statistics which is accurate. The `Optional.of(sourceStats)` is not correct as the output of `PARTIAL` aggregation does not match its input.

Reply:
got it. thanks.